capabilities/writetarget/write_target.go

```diff
@@ -0,0 +1,506 @@
+	GetTransactionStatus(ctx context.Context, transactionID string) (commontypes.TransactionStatus, error)
+}
+
+var (
+	_ capabilities.ExecutableCapability = &writeTarget{}
+)
+
+// chain-agnostic consts
+const (
+	CapabilityName = "write"
+
+	// Input keys
+	// Is this key chain agnostic?
+	KeySignedReport = "signed_report"
+)
+
+type chainService interface {
+	LatestHead(ctx context.Context) (commontypes.Head, error)
+}
+
+type contractReader interface {
+	GetLatestValue(ctx context.Context, readIdentifier string, confidenceLevel primitives.ConfidenceLevel, params, returnVal any) error
+}
+
+type writeTarget struct {
+	capabilities.CapabilityInfo
+
+	config    Config
+	chainInfo monitor.ChainInfo
+
+	lggr logger.Logger
+	// Local beholder client, also hosting the protobuf emitter
+	beholder *monitor.BeholderClient
+
+	cs               chainService
+	configValidateFn func(request capabilities.CapabilityRequest) (string, error)
+
+	nodeAddress      string
+	forwarderAddress string
+
+	targetStrategy       TargetStrategy
+	writeAcceptanceState commontypes.TransactionStatus
+}
+type WriteTargetOpts struct {
+	ID string
+
+	// toml: [<CHAIN>.WriteTargetCap]
+	Config Config
+	// ChainInfo contains the chain information (used as execution context)
+	// TODO: simplify by passing via ChainService.GetChainStatus fn
+	ChainInfo monitor.ChainInfo
+
+	Logger   logger.Logger
+	Beholder *monitor.BeholderClient
+
+	ChainService     chainService
+	ContractReader   contractReader
+	EVMService       commontypes.EVMService
+	ConfigValidateFn func(request capabilities.CapabilityRequest) (string, error)
+
+	NodeAddress      string
+	ForwarderAddress string
+
+	TargetStrategy       TargetStrategy
+	WriteAcceptanceState commontypes.TransactionStatus
+}
+
+// Capability-specific configuration
+type ReqConfig struct {
+	Address   string
+	Processor string
+}
+
+// NewWriteTargetID returns the capability ID for the write target
+func NewWriteTargetID(chainFamilyName, networkName, chainID, version string) (string, error) {
+	// Input args should not be empty
+	if version == "" {
+		return "", fmt.Errorf("version must not be empty")
+	}
+
+	// Network ID: network name is optional, if not provided, use the chain ID
+	networkID := networkName
+	if networkID == "" && chainID == "" {
+		return "", fmt.Errorf("invalid input: networkName or chainID must not be empty")
+	}
+	if networkID == "" || networkID == "unknown" {
+		networkID = chainID
+	}
+
+	// allow for chain family to be empty
+	if chainFamilyName == "" {
+		return fmt.Sprintf("%s_%s@%s", CapabilityName, networkID, version), nil
+	}
+
+	return fmt.Sprintf("%s_%s-%s@%s", CapabilityName, chainFamilyName, networkID, version), nil
+}
+
+// TODO: opts.Config input is not validated for sanity
+func NewWriteTarget(opts WriteTargetOpts) capabilities.ExecutableCapability {
+	capInfo := capabilities.MustNewCapabilityInfo(opts.ID, capabilities.CapabilityTypeTarget, CapabilityName)
```

Comment:
We should probably check the WriteAcceptanceState value. If it's set to Unknown, we should assume that it was not set and override it with Finalized.